src/front/domain_socket.rs

```diff
@@ -42,36 +42,36 @@ impl DomainSocketListener {
     /// - if a file/socket exists at the path specified for the socket and `remove_file`
     /// fails
     /// - if binding to the socket path fails
-    fn init(&mut self) {
-        if cfg!(feature = "systemd-daemon") {
-            // The PARSEC service is socket activated (see parsec.socket file).
-            // systemd creates the PARSEC service giving it an initialised socket as the file
-            // descriptor number 3 (see sd_listen_fds(3) man page).
-            // If an instance of PARSEC compiled with the "systemd-daemon" feature is run directly
-            // instead of by systemd, this call will still work but the next accept call on the
-            // UnixListener will generate a Linux error 9 (Bad file number), as checked below.
-            unsafe {
-                self.listener = Some(UnixListener::from_raw_fd(3));
-            }
-        } else {
-            let socket = Path::new(SOCKET_PATH);
+    pub fn new(timeout: Duration) -> Self {
+        // If this PARSEC instance was socket activated (see the `parsec.socket`
+        // file), the listener will be opened by systemd and passed to the
+        // process.
+        // If PARSEC was service activated or not started under systemd, this
+        // will return `0`.
+        let listener =
+            match sd_notify::listen_fds().expect("Could not retrieve listener from systemd") {
+                0 => {
+                    let socket = Path::new(SOCKET_PATH);
 
-            if socket.exists() {
-                fs::remove_file(&socket).unwrap();
-            }
+                    if socket.exists() {
+                        fs::remove_file(&socket).unwrap();
+                    }
 
-            let listener_val = match UnixListener::bind(SOCKET_PATH) {
-                Ok(listener) => listener,
-                Err(err) => panic!(err),
+                    UnixListener::bind(SOCKET_PATH).expect("Could not bind listen socket")
+                }
+                1 => {
+                    let nfd = sd_notify::SD_LISTEN_FDS_START;
+                    unsafe { UnixListener::from_raw_fd(nfd) }
+                }
+                _ => panic!("Received too many file descriptors"),
             };
 
-            // Set the socket as non-blocking.
-            listener_val
-                .set_nonblocking(true)
-                .expect("Could not set the socket as non-blocking");
```

Comment:
Just FYI, this was inside the non `systemd` case as the socket is set as non-blocking in the `parsec.service` file. Might be good to remove that line [from there](https://github.com/parallaxsecond/parsec/blob/master/systemd-daemon/parsec.service) then.

Reply:
Sorry, I hadn't noticed that. I'll move this to the other branch instead.